File: task-manager/src/app/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Task {
    id: number;
    title: string;
    description: string;
    status: 'TODO' | 'IN_PROGRESS' | 'DONE';
    priority: 'LOW' | 'MEDIUM' | 'HIGH';
    dueDate: string;
    createdAt: string;
    ownerId: number;
    ownerUsername: string;
    assigneeId: number | null;
    assigneeUsername: string | null;
}

export interface TaskRequest {
    title: string;
    description?: string;
    status?: string;
    priority?: string;
    dueDate?: string;
    assigneeId?: number | null;
}

export interface TaskStats {
    total: number;
    todo: number;
    inProgress: number;
    done: number;
}

@Injectable({ providedIn: 'root' })
export class TaskService {
    private baseUrl = 'http://localhost:8080/api/tasks';

    constructor(private http: HttpClient) { }

    getTasks(filters?: { status?: string; priority?: string; search?: string }): Observable<Task[]> {
        let params = new HttpParams();
        if (filters?.status) params = params.set('status', filters.status);
        if (filters?.priority) params = params.set('priority', filters.priority);
        if (filters?.search) params = params.set('search', filters.search);
        return this.http.get<Task[]>(this.baseUrl, { params });
    }

    createTask(task: TaskRequest): Observable<Task> {
        return this.http.post<Task>(this.baseUrl, task);
    }

    updateTask(id: number, task: TaskRequest): Observable<Task> {
        return this.http.put<Task>(`${this.baseUrl}/${id}`, task);
    }

    deleteTask(id: number): Observable<any> {
        return this.http.delete(`${this.baseUrl}/${id}`);
    }

    getStats(): Observable<TaskStats> {
        return this.http.get<TaskStats>(`${this.baseUrl}/stats`);
    }
}

File: task-manager-backend/src/main/java/com/antigravity/taskmanager/service/TaskService.java
package com.antigravity.taskmanager.service;

import com.antigravity.taskmanager.dto.TaskRequest;
import com.antigravity.taskmanager.dto.TaskResponse;
import com.antigravity.taskmanager.model.Task;
import com.antigravity.taskmanager.model.User;
import com.antigravity.taskmanager.repository.TaskRepository;
import com.antigravity.taskmanager.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class TaskService {

    private final TaskRepository taskRepository;
    private final UserRepository userRepository;

    public List<TaskResponse> getTasks(User currentUser, String statusStr, String priorityStr, String search) {
        Task.Status status = statusStr != null ? Task.Status.valueOf(statusStr) : null;
        Task.Priority priority = priorityStr != null ? Task.Priority.valueOf(priorityStr) : null;
        String searchParam = (search != null && !search.isBlank()) ? search : null;

        boolean isAdmin = currentUser.getRole() == User.Role.ADMIN;
        List<Task> tasks;

        if (isAdmin) {
            tasks = taskRepository.findWithFilters(status, priority, searchParam);
        } else {
            tasks = taskRepository.findByOwnerWithFilters(currentUser, status, priority, searchParam);
        }

        return tasks.stream().map(this::toResponse).collect(Collectors.toList());
    }

    public TaskResponse createTask(TaskRequest request, User owner) {
        User assignee = null;
        if (request.getAssigneeId() != null) {
            assignee = userRepository.findById(request.getAssigneeId()).orElse(null);
        }

        Task task = Task.builder()
                .title(request.getTitle())
                .description(request.getDescription())
                .status(request.getStatus() != null ? request.getStatus() : Task.Status.TODO)
                .priority(request.getPriority() != null ? request.getPriority() : Task.Priority.MEDIUM)
                .dueDate(request.getDueDate())
                .owner(owner)
                .assignee(assignee)
                .build();

        return toResponse(taskRepository.save(task));
    }

    public TaskResponse updateTask(Long id, TaskRequest request, User currentUser) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Task not found: " + id));

        boolean isAdmin = currentUser.getRole() == User.Role.ADMIN;
        if (!isAdmin && !task.getOwner().getId().equals(currentUser.getId())) {
            throw new RuntimeException("Access denied: you don't own this task");
        }

        if (request.getTitle() != null) task.setTitle(request.getTitle());
        if (request.getDescription() != null) task.setDescription(request.getDescription());
        if (request.getStatus() != null) task.setStatus(request.getStatus());
        if (request.getPriority() != null) task.setPriority(request.getPriority());
        if (request.getDueDate() != null) task.setDueDate(request.getDueDate());
        if (request.getAssigneeId() != null) {
            userRepository.findById(request.getAssigneeId()).ifPresent(task::setAssignee);
        }

        return toResponse(taskRepository.save(task));
    }

    public void deleteTask(Long id, User currentUser) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Task not found: " + id));

        boolean isAdmin = currentUser.getRole() == User.Role.ADMIN;
        if (!isAdmin && !task.getOwner().getId().equals(currentUser.getId())) {
            throw new RuntimeException("Access denied: you don't own this task");
        }

        taskRepository.delete(task);
    }

    public Map<String, Long> getStats() {
        return Map.of(
                "total", taskRepository.count(),
                "todo", taskRepository.countByStatus(Task.Status.TODO),
                "inProgress", taskRepository.countByStatus(Task.Status.IN_PROGRESS),
                "done", taskRepository.countByStatus(Task.Status.DONE)
        );
    }

    private TaskResponse toResponse(Task task) {
        return TaskResponse.builder()
                .id(task.getId())
                .title(task.getTitle())
                .description(task.getDescription())
                .status(task.getStatus())
                .priority(task.getPriority())
                .dueDate(task.getDueDate())
                .createdAt(task.getCreatedAt())
                .ownerId(task.getOwner().getId())
                .ownerUsername(task.getOwner().getUsername())
                .assigneeId(task.getAssignee() != null ? task.getAssignee().getId() : null)
                .assigneeUsername(task.getAssignee() != null ? task.getAssignee().getUsername() : null)
                .build();
    }
}
